File: app/src/main/java/com/github/malitsplus/shizurunotes/ui/charadetails/SkillAdapter.java
package com.github.malitsplus.shizurunotes.ui.charadetails;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.github.malitsplus.shizurunotes.R;
import com.github.malitsplus.shizurunotes.data.Skill;
import com.github.malitsplus.shizurunotes.databinding.ListItemSkillBinding;

import java.util.ArrayList;
import java.util.List;

public class SkillAdapter extends RecyclerView.Adapter<SkillAdapter.SkillViewHolder> {
    private Context mContext;
    private List<Skill> itemList = new ArrayList<>();

    public SkillAdapter(Context context){
        mContext = context;
    }


    @Override
    @NonNull
    public SkillViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ListItemSkillBinding binding = DataBindingUtil.inflate(
                LayoutInflater.from(parent.getContext()),
                R.layout.list_item_skill, parent, false);

        return new SkillViewHolder(binding);

    }

    //填充每个item的视图
    @Override
    public void onBindViewHolder(SkillViewHolder holder, int position){
        holder.getBinding().setSkill(itemList.get(position));
        holder.getBinding().executePendingBindings();
    }

    @Override
    public int getItemCount(){
        return itemList.size();
    }

    public void update(List<Skill> itemList){
        this.itemList = itemList;
        notifyDataSetChanged();
    }

    public static class SkillViewHolder extends RecyclerView.ViewHolder{

        private ListItemSkillBinding binding;
        SkillViewHolder(ListItemSkillBinding binding){
            super(binding.getRoot());
            this.binding = binding;
        }

        ListItemSkillBinding getBinding(){
            return binding;
        }

    }
}

File: app/src/main/java/com/github/malitsplus/shizurunotes/ui/charadetails/SkillAdapter.kt
package com.github.malitsplus.shizurunotes.ui.charadetails

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.navigation.findNavController
import androidx.recyclerview.widget.RecyclerView
import com.github.malitsplus.shizurunotes.R
import com.github.malitsplus.shizurunotes.data.Skill
import com.github.malitsplus.shizurunotes.databinding.ListItemSkillBinding
import com.github.malitsplus.shizurunotes.ui.shared.SharedViewModelChara
import java.util.*

class SkillAdapter(
    private val sharedChara: SharedViewModelChara
) : RecyclerView.Adapter<SkillAdapter.SkillViewHolder>() {

    var itemList: List<Skill> = ArrayList()

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): SkillViewHolder {
        val binding = DataBindingUtil.inflate<ListItemSkillBinding>(
            LayoutInflater.from(parent.context),
            R.layout.list_item_skill, parent, false
        )
        return SkillViewHolder(
            binding
        )
    }

    //填充每个item的视图
    override fun onBindViewHolder(
        holder: SkillViewHolder,
        position: Int
    ) {
        with(holder.binding){
            skill = itemList[position].also { s ->
                if (s.friendlyMinionList.isNotEmpty()){
                    minionButton.visibility = View.VISIBLE
                    minionButton.setOnClickListener {
                        sharedChara.selectedMinion = s.friendlyMinionList
                        it.findNavController()
                            .navigate(CharaDetailsFragmentDirections
                                .actionNavCharaDetailsToNavMinion()
                            )
                    }
                }
            }
            executePendingBindings()
        }
    }

    override fun getItemCount(): Int {
        return itemList.size
    }

    fun update(itemList: List<Skill>) {
        this.itemList = itemList
        notifyDataSetChanged()
    }

    class SkillViewHolder internal constructor(val binding: ListItemSkillBinding) :
        RecyclerView.ViewHolder(binding.root)

}